### PR TITLE
pkp/pkp-lib#10306 unit tests for queue jobs

### PR DESCRIPTION
**Pull request by touhidurabir**
for 10306

### CODE REVIEW

File: classes/core/PKPApplication.php [hunk 1]
```diff
@@ -319,6 +319,13 @@ public function getUUID(): string
      */
     public function getHttpClient()
     {
+        if (PKPContainer::getInstance()->runningUnitTests()) {
+            $client = Registry::get(\PKP\tests\PKPTestCase::MOCKED_GUZZLE_CLIENT_NAME);
+            if ($client) {
+                return $client;
+            }
+        }
+
```
Comment (touhidurabir) [inline classes/core/PKPApplication.php:328]: part of functionality to mocking ability for guzzel reqeust .

File: classes/notification/NotificationManagerDelegate.php [hunk 2]
```diff
@@ -56,7 +56,7 @@ public function updateNotification(PKPRequest $request, ?array $userIds, int $as
      *
      * @copydoc PKPNotificationOperationManager::createNotification()
      */
-    public function createNotification(PKPRequest $request, ?int $userId = null, ?int $notificationType = null, ?int $contextId = Application::SITE_CONTEXT_ID, ?int $assocType = null, ?int $assocId = null, int $level = Notification::NOTIFICATION_LEVEL_NORMAL, ?array $params = null): ?Notification
+    public function createNotification(?PKPRequest $request = null, ?int $userId = null, ?int $notificationType = null, ?int $contextId = Application::SITE_CONTEXT_ID, ?int $assocType = null, ?int $assocId = null, int $level = Notification::NOTIFICATION_LEVEL_NORMAL, ?array $params = null): ?Notification
```
Comment (touhidurabir) [inline classes/notification/NotificationManagerDelegate.php:59]: do we have any use of `PKPRequest $request` here as it seems this data not used ?

File: tests/PKPTestCase.php [hunk 3]
```diff
@@ -205,8 +220,7 @@ protected function mockRequest($path = 'index/test-page/test-op', $userId = null
         $request->setRouter($router);
 
         // Test user.
-        $session = $request->getSession();
-        $session->setUserId($userId);
+        $request->getSessionGuard()->setUserId($userId);
```
Comment (touhidurabir) [inline tests/PKPTestCase.php:223]: a leftover missing change from https://github.com/pkp/pkp-lib/issues/9566

File: tests/jobs/bulk/BulkEmailSenderTest.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            ->shouldReceive('getCollector')
+            ->withAnyArgs()
+            ->andReturn($userCollectorMock)
+            ->getMock();
+        
+        app()->instance(UserRepository::class, $userRepoMock);
+
+        $bulkEmailSenderJob->handle();
+
+        $this->expectNotToPerformAssertions();
```
Comment (touhidurabir) [inline tests/jobs/bulk/BulkEmailSenderTest.php:75]: Our previous asserting was like `$this->assertNull($bulkEmailSenderJob->handle())` . However the `handle` method return type set as `void` which does turn in null and older PHPUnit version did not have a arg type defined . But the latest PHPUnit version has arg type define as `mixed` which in turn identified as error as return type `void` does not match with arg time `mixed` . Though this does not cause exception but modern IDEs identify it as error . So probably better to go with approach

File: classes/core/PKPContainer.php [hunk 5]
```diff
@@ -590,6 +586,31 @@ public function environment(string ...$environments): string|bool
 
         return $this->get('config')['app']['env'];
     }
+
+    /**
+     * Override Laravel method; always false.
+     * Prevents the undefined method error when the Log Manager tries to determine the driver
+     */
+    public function runningUnitTests(): bool
```
Comment (asmecher) [inline classes/core/PKPContainer.php:594]: This could be removed once we add and configure the Laravel logging service provider, correct? I think this is related to what you're talking about with Erik. It might be worth filing in Github, noting this function in the issue, and adding a FIXME here.

File: classes/notification/NotificationManagerDelegate.php (hunk 2 above)
Comment (asmecher) [reply classes/notification/NotificationManagerDelegate.php:59]: You're right -- this parameter can be removed.

File: tests/PKPTestCase.php [hunk 6]
```diff
@@ -238,6 +253,49 @@ protected function localeToRegExp(string $translation): string
         $escapedPieces = array_map(fn ($piece) => preg_quote($piece, '/'), $pieces);
         return '/^' . implode('.*?', $escapedPieces) . '$/u';
     }
+
+    /**
+     * Mock the mail facade
+     * @see https://laravel.com/docs/10.x/mocking
```
Comment (asmecher) [inline tests/PKPTestCase.php:259]: Better to link to 11, since that's what we're shipping.

File: tests/jobs/doi/DepositSubmissionTest.php [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
...nal for consideration.</li><li>All references have been checked for accuracy and completeness.</li><li>All tables and figures have been numbered and labeled.</li><li>Permission has been obtained to publish all photos, datasets and other material provided with this submission.</li></ul>";s:5:"fr_CA";s:37:"##default.contextSettings.checklist##";}s:15:"uploadFilesHelp";a:2:{s:2:"en";s:249:"<p>Provide any files our editorial team may need to evaluate your submission. In addition to the main work, you may wish to submit data sets, conflict of interest statements, or other supplementary files if these will be helpful for our editors.</p>";s:5:"fr_CA";s:39:"##default.submission.step.uploadFiles##";}}s:20:"_hasLoadableAdapters";b:0;s:27:"_metadataExtractionAdapters";a:0:{}s:25:"_extractionAdaptersLoaded";b:0;s:26:"_metadataInjectionAdapters";a:0:{}s:24:"_injectionAdaptersLoaded";b:0;}s:9:"\0*\0agency";O:43:"APP\\plugins\\generic\\datacite\\DatacitePlugin":4:{s:10:"pluginPath";s:24:"plugins/generic/datacite";s:14:"pluginCategory";s:7:"generic";s:7:"request";N;s:58:"\0APP\\plugins\\generic\\datacite\\DatacitePlugin\0_exportPlugin";N;}s:10:"connection";s:8:"database";s:5:"queue";s:5:"queue";}
```
Comment (asmecher) [inline tests/jobs/doi/DepositSubmissionTest.php:37]: This task in particular has quite a bit of localized content, including some `##untranslated.locale.keys##` (not sure why that's the case -- `default.contextSettings.checklist` is showing untranslated for `fr_CA`, for example, but it does appear in the locale files for OJS, OMP, and OPS). How do you think we're most likely to maintain and regenerate these as needed? If manually, it might be worth trimming out the extraneous localized text into something more minimal.

File: tests/jobs/notifications/NewAnnouncementNotifyUsersTest.php [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    public function testRunSerializedJob(): void
+    {
+        $this->mockMail();
+        
+        $this->mockRequest();
+
+        /** @var NewAnnouncementNotifyUsers $newAnnouncementNotifyUsersJob */
+        $newAnnouncementNotifyUsersJob = unserialize($this->serializedJobData);
+
+        $announcementMock = Mockery::mock(\PKP\announcement\Announcement::class)
```
Comment (asmecher) [inline tests/jobs/notifications/NewAnnouncementNotifyUsersTest.php:60]: This is going to require some work once https://github.com/pkp/pkp-lib/issues/10328 is merged.

File: tests/jobs/notifications/NewAnnouncementNotifyUsersTest.php [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            ->getMock();
+
+        $contextDaoMock = Mockery::mock($contextDaoClass)
+            ->makePartial()
+            ->shouldReceive('getById')
+            ->withAnyArgs()
+            ->andReturn($contextMock)
+            ->getMock();
+
+        DAORegistry::registerDAO(substr(strrchr($contextDaoClass, '\\'), 1), $contextDaoMock);
```
Comment (asmecher) [inline tests/jobs/notifications/NewAnnouncementNotifyUsersTest.php:95]: I think I'd prefer a match statement on e.g. `Application::getName` rather than string parsing and assembly. At some point we'll get rid of `JournalDAO`, for example, and being able to grep for it will be helpful.

**Review by asmecher — CHANGES_REQUESTED** [review]
Just a few minor suggestions -- thanks, @touhidurabir!

File: classes/core/PKPContainer.php (hunk 5 above)
Comment (touhidurabir) [reply classes/core/PKPContainer.php:594]: Not exactly , this has there before this PR and probably will stay . Our Container class extends the laravel core container but We have our very own Application class . Once we able to handle the bootstrapping process to Laravel's foundation Application, we can remove this (as it's reside there) . As few core class required this information, so we have copied it from Laravel's foundation application class to our PKPContainer just to facilitate the dependency.

File: tests/jobs/doi/DepositSubmissionTest.php (hunk 7 above)
Comment (touhidurabir) [reply tests/jobs/doi/DepositSubmissionTest.php:37]: probably we do not need to worry much . First of all, we need set some rules when writing tests for `handle` method of a job . I believe we need to make sure when testing job handle, none of the calls make to real object but mock object only e.g. we need to make sure we have proper mocks for each of the call to classes in `handle` method . 

That saying we can trim down the localized content but do we need to ?

Comment (asmecher) [reply tests/jobs/doi/DepositSubmissionTest.php:37]: Let's not resolve a problem until it becomes one :)
